ecs_ecr: Fix AWS ECR repository creation

deiwin opened the pull request:
##### SUMMARY
Remove the superfluous registry ID from create repository call, because it's not supported.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ecs_ecr

##### ANSIBLE VERSION
```
ansible-playbook 2.4.2.0
  config file = [redacted]
  configured module search path = [u'[redacted]', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
[Boto3 documentation][1] specifies 'repositoryName' as the only expected
argument. The `**build_kwargs(registry_id)` part also adds 'registryId' which,
when executed, fails with: 'Unknown parameter in input: “registryId”, must be
one of: repositoryName'.

[AWS API documentation][2] also lists only the 'repositoryName' parameter. I.e.
this is not a problem with the boto3 library.

The default registry ID for the account that's making the request will be used
when creating the rpository. This means that if the `registry_id` specified by
the user is different from the default registry ID, then the policy changes
following the repository creation would fail, because the repository will have
been created in one repository but subsequent calls try to modify it in
another. Added a safeguard against this scenario.

[1]: https://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.create_repository
[2]: https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_CreateRepository.html

deiwin committed:
ecs_ecr: Remove registry ID from create repository call

[Boto3 documentation][1] specifies 'repositoryName' as the only expected
argument. The `**build_kwargs(registry_id)` part also adds 'registryId' which,
when executed, fails with: 'Unknown parameter in input: “registryId”, must be
one of: repositoryName'.

[AWS API documentation][2] also lists only the 'repositoryName' parameter. I.e.
this is not a problem with the boto3 library.

The default registry ID for the account that's making the request will be used
when creating the rpository. This means that if the `registry_id` specified by
the user is different from the default registry ID, then the policy changes
following the repository creation would fail, because the repository will have
been created in one repository but subsequent calls try to modify it in
another. Added a safeguard against this scenario.

[1]: https://boto3.readthedocs.io/en/latest/reference/services/ecr.html#ECR.Client.create_repository
[2]: https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_CreateRepository.html

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/50336/58/tests

deiwin commented:
Yes, I saw that, but didn't understand why that might be. Would it be possible to check if the same test is passing on `devel`?

mattclay commented:
Yes, it is currently passing in devel.

deiwin committed:
Fix concurrent ECR integration tests

If the `ecr_name` is the same in multiple concurrent test runs, then they can
interfere with one another causing both to fail. The `resource_prefix` is
guaranteed to be unique for different jobs running in CI an so avoids this
issue while also making it easier to identify the test which created the
resource.

deiwin commented:
After boring through the test logs, I think I understand why those tests failed. It's a race condition between the python 2.7 and 3.6 test runs (`#57` and `#58` respectively).

Here's what happened (at least in the retries):
- Both tests set the repository name with `ecr-test-{{ ansible_date_time.epoch }}`, which evaluates to "ecr-test-1515758673" for both
- Test `#57` creates the repository at 12:04:35
- Test `#58` also tries to create the repository at 12:04:35, but fails because it already exists
- Test `#58` deletes the repository at 12:04:36, because of the failure (the _always_ step of the test)
- Test `#57` tries to check that `changed` is False when re-creating, but fails, because `#58` has now deleted the repository and so the test creates it again instead.

Added https://github.com/ansible/ansible/pull/34798/commits/95aa8d6fe676c0c45fb1d737bd1a7a8518bf0f3f to avoid this issue.

deiwin commented:
Thanks for the pointer, @mattclay. Updated the last commit.

willthames commented:
Thanks @deiwin
